source/agora/consensus/ValidatorSet.d

```diff
@@ -546,6 +547,44 @@ public class ValidatorSet
 
         return true;
     }
+
+    /***************************************************************************
+
+        Restore validators' information from blocks
+
+        Params:
+            blocks = blocks to get validators' information
+            finder = the delegate to find UTXOs with
+
+    ***************************************************************************/
+
+    public void restoreValidators (const(Block)[] blocks,
+        scope UTXOFinder finder) @safe nothrow
```

Comment:
This function is not called from anywhere, how do you plan to have it used ?
Also passing a full array seems quite wasteful. A range might be more appropriate, but it's hard to say without seeing the usage. Do you have a WIP for it ? I understand if you want to keep it separate from this PR.

Reply:
> This function is not called from anywhere, how do you plan to have it used ?

It was for smaller PR. But, It would be good to add the code calling this function. I will add more commits for that.

> Also passing a full array seems quite wasteful. A range might be more appropriate,

Please explain more about how I can pass a range? In current plan, this function will be called from Node class through the EnrollmentManger.

Reply:
In order to receive a range:
```D
    public void restoreValidators (R) (R blocks, scope UTXOFinder finder) @safe nothrow
        if (isInputRange!R)
```

If you want to ensure that only blocks are passed, you can do:
```D
        if (isInputRange!R && is(ElementType!R : const(Block))
```

Reply:
> public void restoreValidators (R) (R blocks, scope UTXOFinder finder) @safe nothrow
>         if (isInputRange!R)

Thank you for the information. I have read the doc you let me know yesterday and it really helped me understand the concept of the Range. But, I have decided to use `IBlockStorage` instead of `InputRange` because only `ValidatorSet` knows how many blocks are needed in order to restore validators' information in the headers like as follows.
```
public void restoreValidators2 (ulong last_height, IBlockStorage storage,
        scope UTXOFinder finder) @safe
    {
        Block block;
        foreach (height; last_height - ValidatorCycle .. last_height)
```

Reply:
I'm not sure I follow. Ranges are an higher order data-structure, so they should work for that use case too.
For example, if you want to repeat an action X times:

```D
iota(X).map!(_ => my_action_here);
```

So in pseudo code:
```
iota(X).map!(o => storage.getHeight(last_height - ValidatorCycle + o)).map!(block => block.header.enrollments)
    .filter!(enroll => !this.hasEnrollment(enroll.utxo_key)).each!(enroll => this.add(block.header.height, finder, enroll));
```

In reality it's a tiny bit more complicated because you need to flatten the `enrollments` range and keep `block` around for the last `each` call, but there's hardly a concept you can't represent with `Range`. Let me know when you have implemented the `IBlockStorage` approach and I'll give it another review.

Reply:
Sorry for my incomplete example. I'll request another review with `IBlockStorage` and get your code above understood.

Reply:
I think it'll help if you watch this next time.
Ledger.getBlocksFrom (size_t start_height)

Reply:
> I think it'll help if you watch this next time.
Ledger.getBlocksFrom (size_t start_height)

Thank you for the information.